src/main/scala/tech/cryptonomic/conseil/generic/rpc/DataFetcher.scala

```diff
@@ -0,0 +1,189 @@
+      of encoded data (e.g. json), paired with its input value (to preserve correlation)
+  * - decodeData function that will decode every single endoded value to a full-blown output scala data type
+  *
+  * The effectul function is defined in term of a "Kleisli arrow", which is nothing but a fancy-named wrapper for A => F[B], where F[_] is the so-called effect
+  * Using Kleisli enables a plethora of useful methods to call, that will more easily convert a Kleisli arrow to something that accepts different input, or that
+  * adds further operations on input or output, or that adds other effects, like traversing or folding a collection of results if the B result type allows that.
+  *
+  * This enables us to easily compose the `fetchData` function and `decodeData` function to a general `fetch` function, that will use the former to retrieve a list of
+  * encoded values, paired with input, and apply the decoding to each, all the while preserving the correlation with the original input value.
+  *
+  * In addition to this, we might want to apply multiple decoding steps in parallel, and pair the decoded results with the same inputs
+  * Or still we might want to apply more than one fetcher to the same set of inputs, and pair all the results, in a single pass
+  * This is where the `addDecoding` and `mergeResults`/`fetchCombine` combinators comes useful.
+  */
+object DataFetcher {
+  import cats.syntax.applicativeError._
+  import cats.syntax.semigroupal._
+
+  /** using the combination of the provided fetcher functions we fetch all data with this general outline:
+    * 1. get encoded (e.g. json) values from the node, given a traversable input collection
+    * 2. keep the correlation with the input and decode each value (e.g. from json) to a corresponding output
+    * 3. return the traversable collection containing paired input and output values, all wrapped in the effect F
+    * The resulting effect will also encode a possible failure in virtue of the implicit MonadError instance that is provided
+    */
+  def getFetcher[Eff[_], In, Out, Err](
+    implicit fetcher: Aux[Eff, In, Out, _, Err],
+    flatMapper: FlatMap[Eff],
+    errorHandling: ApplicativeError[Eff, Err]
+  ): Kleisli[Eff, In, Out] =
+    fetcher.fetchData.onError { case err => Kleisli.pure(fetcher.onDataFetchError(err)) }
+      .andThen(
+        fetcher.decodeData.onError { case err => Kleisli.pure(fetcher.onDecodingError(err)) }
+      )
+
+  /* An alias used to define function constraints on the internal dependent types (i.e. `I`, `O`, `E`)
+   * which would be otherwise opaque in any function signature
+   *
+   * Example: the fetcher for tezos blocks might have type `Aux[Future, List, Throwable, Int, BlockData, String]
+   * where the Int is for the offset from a given block, and String is a representation of a json value.
+   */
+  type Aux[Eff[_], In, Out, Encoding, Err] = DataFetcher[Eff, In, Out] {
```

Comment:
Aux is an established name, having become the name for what is well known as the "Aux Pattern".

Std was a shortcut name I tried to find to imply the commonly used form of parameterization. It's defined so that method type paramters' constraints could be expressed easily.
Any name that conveys the same meaning - i.e. that of being the standardly-used version - will work fine.

Reply:
Could we rename `Std` to something like `StandardDataFetcher`? It makes the code a lot more readable. People might be put off my how abstract it seems.